From 03a4e55fbb4d017f245521125df46124d0eb9bf8 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Fri, 12 Jan 2024 08:53:34 +0100
Subject: [PATCH 0756/1141] irqchip/stm32mp-exti: fix irq type propagation on
 configurable events

EXTI configurable events are only able to generate a level high
interrupt for the parent interrupt controller (typically the GIC).

In set_type(), check the event type and propagate the irq type for
direct events only, while force level high type for configurable
events.

This fixes a missing interrupt during USB wake-up on STM32MP25.

Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Reported-by: Pankaj Dev <pankaj.dev@st.com>
Change-Id: Iba87605f5143d36d088f436c8e697d8f3885b0e4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/351763
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Alexandre TORGUE <alexandre.torgue@foss.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/irqchip/irq-stm32mp-exti.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-stm32mp-exti.c b/drivers/irqchip/irq-stm32mp-exti.c
index 6f7cc8565062..3bf33f52219c 100644
--- a/drivers/irqchip/irq-stm32mp-exti.c
+++ b/drivers/irqchip/irq-stm32mp-exti.c
@@ -416,8 +416,9 @@ static int stm32mp_exti_set_type(struct irq_data *d, unsigned int type)
 unlock:
 	raw_spin_unlock(&chip_data->rlock);
 
+	/* configurable events only propagate level high type to GIC */
 	if (d->parent_data->chip)
-		irq_chip_set_type_parent(d, type);
+		irq_chip_set_type_parent(d, IRQ_TYPE_LEVEL_HIGH);
 
 	return err;
 }
-- 
2.39.2

